### ISSUE TITLE
Request : need x11vnc in the docker containers

### ISSUE THREAD

**Issue by abhagupta** [sitespeedio/sitespeed.io-docker#8]
@soulgalore . Is it possible to get an updated docker image for chrome which has `x11vnc`? 
I think chrome is having an issue in which it says "Please start google chrome as a normal user", which I found when I recreated `sitespeed.io-chrome` docker (https://github.com/sitespeedio/sitespeed.io-docker/blob/master/browsers/chrome/Dockerfile)  and installed x11vnc.  x11vnc helped me visualize the chrome inside the docker container.  The fix for the above issue was to pass `google-chrome --user-data-dir` in the command. 

Now I don't want to use my customized Dockerfile and miss out any other goodies that you may add.  

If this request is not possible or is going to take too much time, can you suggest me any other solution on how to pass in `--user-data-dir` to chrome command?  Or see the XVFB running inside the current running container?

I think installing x11vnc would be a good idea anyway... :)

**Comment by soulgalore** [comment]
Hey @abhagupta , yes I can add it later today and rebuild. Do you have time to do a PR for the docs how you can use it (https://github.com/sitespeedio/sitespeed.io/tree/master/docs/documentation/sitespeed.io/docker)? It's a good idea to be able to check whats happening. We are releasing 4.0 today/tomorrow if everything works out so I'll add it only to the new structure.

About the error. Let me check that, maybe we can fix that in container, ping @tobli .

Best
Peter

**Comment by soulgalore** [comment]
Added it https://github.com/sitespeedio/docker-browsers/commit/d7d29b686ebdb023ad5b4f7bf7acbeff8319481c  but haven't rebuilt the chain, doing that later today .

**Comment by abhagupta** [comment]
Thanks @soulgalore . I am going to test today. I have created PR to add docs for using `x11vnc` . https://github.com/sitespeedio/sitespeed.io/pull/1274 . Please feel free to modify/correct/add and put it at the right place.

**Comment by abhagupta** [comment]
@soulgalore  unfortunately, the final sitespeed io  image that I built using tsitespeedio/docker-browsers@d7d29b6 didnt work for me. It exits right after I see the message 

```
Fri Oct 28 00:04:28 UTC 2016
Google Chrome 54.0.2840.71 
```

  So I am waiting for your published image. I also noticed that you are sending `--user-data-dir` option in browsertime. So that may not be needed. 

However, while troubleshooting with sitespeedio/sitespeed.io docker and manually adding x11vnc, found that  there is another option `--no-default-browser-check` which might be needed for chrome to open without waiting for user to click OK when message prompts.   Is there an easy way to send the commands to browsertime?

**Comment by soulgalore** [comment]
yes I know I didn't push that image yesterday, can do patch release adding the flags to Chrome.

**Comment by soulgalore** [comment]
Checked and we already use "--no-default-browser-check" lets see if there's another flag.

**Comment by abhagupta** [comment]
@soulgalore . I have upgraded to docker image `sitespeed.io:4.0-alpha`. Noticed that you do not support `--btConfig` option anymore. Is `--config` the new option to send options to `browsertime`?  Chrome is not opening anymore for me.

I tried `--config`, and I see only these options sent to chrome. 

```
/opt/google/chrome/chrome 
--allow-running-insecure-content 
--disable-add-to-shelf 
--disable-background-downloads 
--disable-background-networking
 --disable-bundled-ppapi-flash 
--disable-client-side-phishing-detection 
--disable-component-update 
--disable-datasaver-prompt 
--disable-default-apps
 --disable-desktop-notifications 
--disable-domain-reliability
 --disable-hang-monitor
 --disable-plugins-discovery 
--disable-popup-blocking 
--disable-prompt-on-repost 
--disable-save-password-bubble 
--disable-sync 
--disable-translate -
-disable-web-resources 
--enable-experimental-extension-apis 
--enable-logging 
--ignore-certificate-errors 
--load-extension=/tmp/.org.chromium.Chromium.VhJwVS/internal 
--log-level=0 
--metrics-recording-only 
--new-window --no-default-browser-check 
--no-first-run 
--password-store=basic --process-per-tab 
--remote-debugging-port=12891 
--safebrowsing-disable-auto-update 
--test-type=webdriver 
--use-mock-keychain --user-data-dir=/tmp/.org.chromium.Chromium.3WCfjj data:,
```

My command 
`docker run  -t  -v "$(pwd)":/sitespeed.io  -v "$(pwd)/tmp/chrome.json:/tmp/chrome.json" sitespeedio/sitespeed.io:4.0-alpha sitespeed.io https://www.google.com --noYslow true -d 0 -b chrome --config /tmp/chrome.json`

Also, checked that the chrome.json has correct data (which used to work in sitespeed version 3)

```
{
    "noSandbox": true
}
```

**Comment by soulgalore** [comment]
Hey @abhagupta you should be able to use 4.0.0 tag (but its still missing that extra dependency). in the new version you can pass all options to browsertime using the cli. If run help for browsertime and take that param name and prepend with browsertime. and send it to sitespeed.io and it should work. I can make a better example when i'm in front of the computer. I've checked for the parameter in the source for browsertime, will check again then.

Best
Peter

**Comment by soulgalore** [comment]
You should be able to try it out with https://hub.docker.com/r/sitespeedio/sitespeed.io-autobuild/ (it autobuilds per checkin) so it will have _x11vnc_.

About _--no-default-browser-check_  send it in 4.0.0 and here's the complete list: https://github.com/sitespeedio/browsertime/blob/master/lib/core/webdriver/chrome.js#L17

There's a lot that changed with params for 4.0. The full options: https://www.sitespeed.io/documentation/sitespeed.io/configuration/#the-options

But then also we pass all parameters to other plugins. So if you have browsertime installed/checked out and run help you get all options:

<pre>
bin/browsertime.js [options] <url>
timeouts
  --timeouts.browserStart       Timeout when waiting for browser to start, in
                                milliseconds           [number] [default: 60000]
  --timeouts.pageLoad           Timeout when waiting for url to load, in
                                milliseconds          [number] [default: 300000]
  --timeouts.script             Timeout when running browser scripts, in
                                milliseconds           [number] [default: 80000]
  --timeouts.pageCompleteCheck  Timeout when waiting for page to complete
                                loading, in milliseconds
                                                      [number] [default: 300000]
chrome
  --chrome.args                        Extra command line args to pass to the
                                       chrome process (e.g. --no-sandbox)
  --chrome.binaryPath                  Path to custom Chrome binary (e.g. Chrome
                                       Canary). On OS X, the path should be to
                                       the binary inside the app bundle, e.g.
                                       /Applications/Google Chrome
                                       Canary.app/Contents/MacOS/Google Chrome
                                       Canary
  --chrome.mobileEmulation.deviceName  Name of device to emulate. Works only
                                       standalone (see list in Chrome DevTools,
                                       but add company like 'Apple iPhone 6')
  --chrome.mobileEmulation.width       Width in pixels of emulated mobile screen
                                       (e.g. 360)                       [number]
  --chrome.mobileEmulation.height      Height in pixels of emulated mobile
                                       screen (e.g. 640)                [number]
  --chrome.mobileEmulation.pixelRatio  Pixel ratio of emulated mobile screen
                                       (e.g. 2.0)
  --chrome.android.package             Run Chrome on your Android device. Set to
                                       com.android.chrome for default Chrome
                                       version.
  --chrome.android.deviceSerial        Choose which device to use. If you do not
                                       set it, random device will be used.
firefox
  --firefox.binaryPath  Path to custom Firefox binary (e.g. Firefox Nightly). On
                        OS X, the path should be to the binary inside the app
                        bundle, e.g.
                        /Applications/Firefox.app/Contents/MacOS/firefox-bin
selenium
  --selenium.url  URL to a running Selenium server (e.g. to run a browser on
                  another machine).
experimental
  --experimental.dumpChromePerflog  Dump Chromes performance log with Network.*
                                    events to disk.
  --experimental.video              Record a video and collect SpeedIndex. Needs
                                    FFMPEG & Python dependencies for
                                    VisualMetrics
connectivity
  --connectivity.profile, -c     The connectivity profile. Default connectivity
                                 engine is tsproxy
          [choices: "3g", "3gfast", "3gslow", "2g", "cable", "native", "custom"]
                                                             [default: "native"]
  --connectivity.downstreamKbps  This option requires --connectivity.profile be
                                 set to "custom".
  --connectivity.upstreamKbps    This option requires --connectivity.profile be
                                 set to "custom".
  --connectivity.latency         This option requires --connectivity.profile be
                                 set to "custom".
  --connectivity.tsproxy.port    The port used for ts proxy      [default: 1080]
  --connectivity.tc.device       The connectivity device. Used for engine tc.
                                                               [default: "eth0"]
  --connectivity.engine          The engine for connectivity. Tsproxy needs
                                 Python 2.7. TC needs tc, modprobe and ip
                                 installed to work. Running tc inside Docker
                                 needs modprobe to run outside the container.
                                 [choices: "tc", "tsproxy"] [default: "tsproxy"]
Options:
  --browser, -b        Specify browser
                              [choices: "chrome", "firefox"] [default: "chrome"]
  --screenshot         Save one screen shot per iteration.             [boolean]
  --pageCompleteCheck  Javascript snippet is repeatedly queried to see if page
                       has completed loading (indicated by the script returning
                       true).
  --iterations, -n     Number of times to test the url (restarting the browser
                       between each test)                  [number] [default: 3]
  --prettyPrint        Enable to print json/har with spaces and indentation.
                       Larger files, but easier on the eye.
                                                      [boolean] [default: false]
  --delay              Delay between runs, in milliseconds [number] [default: 0]
  --preScript          Task(s) to run before you test your URL (use it for login
                       etc). Note that --preScript can be passed multiple times.
  --postScript         Task(s) to run after you test your URL (use it for logout
                       etc). Note that --postScript can be passed multiple
                       times.
  --script             Add custom Javascript to run on page. If a single js file
                       is specified, it will be included in the category named
                       "custom" in the output json. Pass a folder to include all
                       .js scripts in the folder, and have the folder name be
                       the category. Note that --script can be passed multiple
                       times.
  --userAgent          Override user agent
  --silent, -q         Only output info in the logs, not to the console. Enter
                       twice to suppress summary line.                   [count]
  --output, -o         Specify file name for Browsertime data (ex:
                       'browsertime'). Unless specified, file will be named
                       browsertime.json
  --har                Specify file name for .har file (ex: 'browsertime').
                       Unless specified, file will be named browsertime.har
  --skipHar            Pass --skipHar to not collect a HAR file.       [boolean]
  --config             Path to JSON config file
  --viewPort           Size of browser window WIDTHxHEIGHT
  --resultDir          Set result directory for the files produced by
                       Browsertime
  -h, --help           Show help                                       [boolean]
  -V, --version        Show version number                             [boolean]
</pre>

And then in sitespeed just pre-pend browsertime. So if I want to change which chrome to use <code>--chrome.binaryPath</code> you in sitespeed.io use <code>--browsertime.chrome.binaryPath</code>.

Also for 4.0 we automatically run sitespeed.io in Docker so the params is changed, you can check the docs here:
https://www.sitespeed.io/documentation/sitespeed.io/docker/

Best
Peter

**Comment by abhagupta** [comment]
Thanks @soulgalore . This helps a lot :).  I could run chrome with version sitespeed version 4 using following commmand 

`docker run  -t  -v "$(pwd)":/sitespeed.io sitespeedio/sitespeed.io:4.0-alpha sitespeed.io  https://www.google.com --noYslow true -d 0 -b chrome --browsertime.chrome.args no-sandbox`

I am going to try the new docker image you built. I think sitespeed 4 might have fixed my original issues of chrome timing out anyway..   Thanks again!  You can close this ticket if you need.

**Comment by tobli** [comment]
Great to hear! FYI, in 4.0 you can skip --noYslow (since we don't use YSlow anymore) and -d 0 (since we don't crawl by default). Chrome is the default browser, so -b chrome is technically not needed, although it can be nice to be explicit.

**Comment by soulgalore** [comment]
Just checking @abhagupta does this work now as you expected or do you still get "Please start google chrome as a normal user"?

**Comment by abhagupta** [comment]
Yes @soulgalore . the problem has been resolved with upgrade to sitepseed v4.

**Comment by abhagupta** [comment]
But presence of X11vnc is still needed :-) . Thanks for following up!

**Comment by soulgalore** [comment]
yes, that should have been included in 4.0.2?
